### PR TITLE
CRM-20144 Fix contact.get api break where groups no longer returned i…

### PR DESCRIPTION
**Pull request by seamuslee001**
…f using title in params

### CODE REVIEW

**Comment by seamuslee001** [comment]
@eileenmcnaughton @adixon This seems to fix the issue, Eileen does this look sensible plus this adds a test  as well

**Comment by seamuslee001** [comment]
sorry pinged wrong Allan @twomice

**Comment by twomice** [comment]
@seamuslee001 Looks like this Is related to https://issues.civicrm.org/jira/browse/CRM-20144, yeah? Thanks for working on this!  If there's a way to link the PR and the issue, it will probably be good for posterity.

**Comment by eileenmcnaughton** [comment]
Actually @seamuslee001 I think the handling should be at the api wrapper not in the BAO layer & we should probably use the Group pseudoconstant fn

**Comment by seamuslee001** [comment]
@eileenmcnaughton I'm not sure how that is easier than handling at the BAO layer

1. Handling at the API level mans that we need to handle the situations whereby we might be in groups = <title> or groups = array('IN" => array(<title> => 1, <title2> => 2))) etc

Also i'm not sure where in the api layer we should be trying to handle this

**Comment by eileenmcnaughton** [comment]
The issue re where boils down to 'don't support api legacy formats in the BAO layer'

Re the how - a couple of thoughts:
1) the identified broken legacy format is 'group' => "Administrators". I'm not clear that this extra level of nastiness should be in scope ; groups = array('IN" => array(<title> => 1, => 2))) 
2) we might be able to use existing api mechanisms - ie. declare the field on the metadata & add the pseudocontant details
3) this is where the api layer does the wrangling https://github.com/eileenmcnaughton/civicrm-core/blob/5147800ea30caf57b6de80004dca53fa00a589b9/api/v3/utils.php#L2234-L2234 
4) worst case jam a civicrm_apiv3_contact_legacy_handling_of_group_field function on the contact api

**Comment by eileenmcnaughton** [comment]
hmm - perhaps the priority is to ensure this DOES work:

IN => array(1,5,7)

so at least we definitely support the formats we LIKE

**Comment by seamuslee001** [comment]
@eileenmcnaughton I have in the latest commit tried moving the special handling into the API layer i think it will be ok can you take a look

File: api/v3/Contact.php [hunk 1]
```diff
@@ -376,6 +376,30 @@ function _civicrm_api3_contact_get_supportanomalies(&$params, &$options) {
     unset($params['filter.group_id']);
     $options['input_params']['group'] = $groups;
   }
+  if (isset($params['group'])) {
+    $groups = $params['group'];
+    $allGroups = CRM_Core_PseudoConstant::group();
+    if (is_array($groups) && in_array(key($groups), CRM_Core_DAO::acceptedSQLOperators(), TRUE)) {
+      $groupsArray = $groups[key($groups)];
+      foreach ($groupsArray as $group => &$ignore) {
```
Comment (eileenmcnaughton) [inline api/v3/Contact.php]: I don't like this stylistically - assigned unrequired vars - esp as a reference - I would use array_keys either when forming or when iterating the $groupsArray var

File: api/v3/Contact.php [hunk 2]
```diff
@@ -376,6 +376,30 @@ function _civicrm_api3_contact_get_supportanomalies(&$params, &$options) {
...
+        if (!is_numeric($group) && array_search($group, $allGroups)) {
+          $group = array_search($group, $allGroups);
+        }
+      }
+      $groups[key($groups)] = $groupsArray;
+    }
+    elseif (is_array($groups)) {
+      foreach ($groups as $k => &$group) {
+        if (array_search($group, $allGroups)) {
+          $group = array_search($group, $allGroups);
```
Comment (eileenmcnaughton) [inline api/v3/Contact.php]: I can't see this being used anywhere

File: tests/phpunit/api/v3/ContactTest.php [hunk 3]
```diff
@@ -3166,4 +3166,47 @@ public function testContactGetReturnGroup() {
     }
   }
 
+  /**
+   * CRM-20144 Verify that passing title of group works as well as id
+   */
```
Comment (eileenmcnaughton) [inline tests/phpunit/api/v3/ContactTest.php]: Perhaps comment on the tested formats in here too for clarity

**Comment by eileenmcnaughton** [comment]
@samuelsov I added a few comments. This is confusing so I feel like it needs extra commenting for people who come later

**Comment by twomice** [comment]
@eileenmcnaughton Was that last comment for Seaumus instead of Samuel?

**Comment by eileenmcnaughton** [comment]
yes......

File: api/v3/Contact.php (hunk 2 above)
Comment (seamuslee001) [inline api/v3/Contact.php]: when you say @eileenmcnaughton when you say  this not being used what do you mean?

File: api/v3/Contact.php (hunk 1 above)
Comment (seamuslee001) [inline api/v3/Contact.php]: ok will look and see what i can do

**Comment by seamuslee001** [comment]
@twomice @eileenmcnaughton would you mind reviewing this again, I now have 3 sets of tests testing 3 forms of the param and i think its pretty good now would appreciate your feelings on it

**Comment by eileenmcnaughton** [comment]
I think it looks OK - there are possibly some formats we might have got away without supporting that we now are locking in, but I think by supporting & converting them at the api layer the code cost is low. We should resolve to only recommend one format - ie

'group' => array('IN' => array(1,4,5));
'group' => array('IN' => array('bob','monkeys'));

I know that looks like 2 formats but our standard is that we are interchangeable about ids vs resolved pseudoconstants.

We should possibly replace the example 'GroupFilterUsingContactApi' with something with one of those.

(I don't think examples  / documentation are a blocker to merging - so giving merge on pass - but would be good to sort)

**Comment by seamuslee001** [comment]
@eileenmcnaughton good to merge? I'll look at updating the examples

**Comment by eileenmcnaughton** [comment]
Thanks for this @seamuslee001 

@colemanw @seanmadsen what does the explorer / documentation promote? I think we should prefer these
civicrm_api3('Contact', 'get', array(
'group' => array('IN' => array(1,4,5))
);

Or
'group' => array('IN' => array('bob','monkeys'));

(which are synonymous from the api POV)

**Pull request merged**